### PR TITLE
Use fused_batch_norm in LayerNormalization

### PR DESCRIPTION
**Pull request by kaixih**
Current LayerNormalization uses nn.batch_normalization, which might be not as efficient as nn.fused_batch_norm.

This PR applies some reshaping on the inputs and uses nn.fused_batch_norm for better performance.

fyi @nluehr

### CODE REVIEW

File: tensorflow/python/keras/layers/normalization.py [hunk 1]
```diff
@@ -945,6 +949,19 @@ def __init__(self,
...
+    self.fused = fused
+
+  def _raise_if_fused_cannot_be_used(self):
+    """Raises a ValueError if fused implementation cannot be used.
+
+    Check if the axis is contiguous. It is possible to squeeze the dims to
+    NCHW-friendly when the axis is contiguous.
+    """
+    axis = sorted(self.axis)
+    for i in range(1, len(axis)):
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: Can u just check if axis[-1] - axis[0] == len(axis) - 1?

File: tensorflow/python/keras/layers/normalization.py [hunk 2]
```diff
@@ -997,29 +1017,67 @@ def call(self, inputs):
...
+      # Compute layer normalization using the batch_normalization function.
+      outputs = nn.batch_normalization(
+          inputs,
+          mean,
+          variance,
+          offset=offset,
+          scale=scale,
+          variance_epsilon=self.epsilon)
+    else:
+      # Squeeze self.axis to C in NCHW of squeezed_shape
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: I am very curious about the NCHW here. If I understand it correctly, the layerNorm can be used for all the format of the input data, regardless whether its from conv layer or not, whereas the NCHW/NHWC is only for conv layer data format. Why it is involved here?

File: tensorflow/python/keras/layers/normalization.py [hunk 3]
```diff
@@ -997,29 +1017,67 @@ def call(self, inputs):
...
+          mean,
+          variance,
+          offset=offset,
+          scale=scale,
+          variance_epsilon=self.epsilon)
+    else:
+      # Squeeze self.axis to C in NCHW of squeezed_shape 
+      squeezed_shape = [1] * 4
+      axis = sorted(self.axis)
+      for dim in range(0, ndims):
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: Can u add some comment here to explain the details? If I understand it correctly, any dim before axis is compressed to N, any dim within axis is compressed to C, and dim after axis is compressed H, and W is not used?

File: tensorflow/python/keras/layers/normalization.py [hunk 4]
```diff
@@ -997,29 +1017,67 @@ def call(self, inputs):
...
+                               squeezed_shape[0] != -1 else -1)
+        elif dim in axis:
+          squeezed_shape[1] *= dim_val
+        else:
+          squeezed_shape[2] = ((squeezed_shape[2] * dim_val) if dim_val and
+                               squeezed_shape[2] != -1 else -1)
+
+      axis_dim_val = squeezed_shape[1]
+      squeezed_shape = ops.convert_to_tensor(squeezed_shape)
+      inputs = array_ops.reshape(inputs, squeezed_shape)
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: For reshape, I think at most of the dim can be -1. What if both N and H are -1 here? Will to cause any issue?

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Can u also add a unit test to compare the numerical result between fused and non-fused layer, and make sure they are same?

**Comment by houtoms** [comment]
Hi @qlzh727  Thanks for your comments. I found a logic error in previous commits. This new commit fixes it. I also made some changes and added a new test based on your comments. Please check.

Basically, this PR tries to collapse the given axis params (the dims that is gonna be normalized over) to (HW) in NCHW and other axis to C (because fused_batch_norm will not normalize the C axis).

Also, the current PR only supports the case that given axis can be collapsed to the last axis, meaning we can put the collapsed axis to (HW) in NCHW without doing a transpose op.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: tensorflow/python/keras/layers/normalization.py [hunk 5]
```diff
@@ -945,6 +950,25 @@ def __init__(self,
...
+    self.fused = fused
+
+  def _raise_if_fused_cannot_be_used(self, ndims):
+    """Raises a ValueError if fused implementation cannot be used.
+
+    Check if the axis is contiguous and can be collapsed into the last axis.
+    """
+    axis = sorted(self.axis)
+    err = False
+    if axis[-1] - axis[0] != len(axis) - 1:
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: I would swap these two if-statements, since what's currently the first if-statement doesn't make much sense unless you know about the second. And, add a comment this tests if the axis is contiguous, since you know axis does not have duplicate values.

Or, just use an explicit for-loop, to assert each axis element is the correct value. Performance isn't a concern here.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/layers/normalization.py [hunk 6]
```diff
@@ -1010,16 +1034,60 @@ def _broadcast(v):
...
+      def _set_const_tensor(val, dtype, shape):
+        return array_ops.fill(shape, constant_op.constant(val, dtype=dtype))
+
+      scale = _set_const_tensor(1.0, inputs.dtype, [pre_dim])
+      offset = _set_const_tensor(0.0, inputs.dtype, [pre_dim])
+
+      # Compute layer normalization using the fused_batch_norm function.
+      outputs, _, _ = nn.fused_batch_norm(
+          inputs,
+          scale=scale,
```
[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/normalization.py]: I don't think we can reshape gamma and beta as input to the fused_batch_norm. Basically, batch norm requires mean/var has the same shape with gamma/beta. Suppose the input the NxD and the batch norm is over N, both mean/var and gamma/beta are 1xD. However, by definition, the layer norm is over D and the mean/var is Nx1 but gamma/beta is 1xD. To utilize the fused batch norm in the layer norm, we let the fused batch norm to normalize over D (instead of N), which will require mean/var and gamma/beta in shape of Nx1. However, our gamma/beta is 1xD and we cannot reshape it to Nx1. So, our strategy is to give a Nx1 const gamma/beta (1, 0) to do a "redundant" scale and offset. And then separately, we handle the meaningful scale and offset using the real 1xD gamma/beta.

File: tensorflow/python/keras/layers/normalization_test.py [hunk 7]
```diff
@@ -515,86 +516,171 @@ def _run_layernorm_correctness_test(layer, dtype='float32'):
...
-    testing_utils.layer_test(
-        keras.layers.LayerNormalization,
-        kwargs={'scale': False,
-                'center': False},
-        input_shape=(3, 3))
+    devices = ["/cpu:0"]
+    fused_vals = [False]
+    if tf_test_util.is_gpu_available():
+      devices.append("/gpu:0")
+      fused_vals.append(True)
```
[the remark this answers is not in the document]
Comment (houtoms) [reply tensorflow/python/keras/layers/normalization_test.py]: The CPU implementation only has NHWC support. But in our case, we have to form the layer form problem to NHWC. So, GPU is our only choice.

**Comment by houtoms** [comment]
Hi @reedwm Thanks for your comments. I made some additional changes according to your suggestion. Please check.

File: tensorflow/python/keras/layers/normalization.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply tensorflow/python/keras/layers/normalization.py]: Can you add a comment? Something like "self.gamma and self.beta have the wrong shape for fused_batch_norm, so we cannot pass them as the scale and offset parameters."

File: tensorflow/python/keras/layers/normalization_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (reedwm) [reply tensorflow/python/keras/layers/normalization_test.py]: Can you put this logic into a separate function then, that returns a list of (device, fused) pairs? Then you can iterate over with "for device, fused in self._get_test_configs()".

Also maybe test nonfused with GPU as well, by adding a third (device, fused) pair.

Also add a comment somewhere in the non-test file saying only GPU is supported when fused=True, and a TODO to add CPU support.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by houtoms** [comment]
@reedwm Thx. I have made the changes accordingly. Please have a look.

**Review by reedwm — APPROVED** [review]

File: tensorflow/python/keras/layers/normalization.py [hunk 8]
```diff
@@ -922,6 +926,7 @@ def __init__(self,
                beta_constraint=None,
                gamma_constraint=None,
                trainable=True,
+               fused=False,
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: Please put "fused" after "name". New arguments must always be added to the end of the argument list.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by houtoms** [comment]
@reedwm Done.

**Review by reedwm — DISMISSED** [review]

**Comment by houtoms** [comment]
Hi @reedwm There are still four tests failing, but I feel the failure is not related to this PR. Could you please help check?

**Comment by reedwm** [comment]
Looking into it. This should be merged soon.

**Comment by reedwm** [comment]
Hi @houtoms,

As discussed offline, there is some reluctance to expose the "fused" parameter. Ideally, the implementation would use the fused version if possible, assuming there are no significant numeric differences.

Unfortunately, as you mentioned, FusedBatchNorm does not support NCHW on the CPU, which means using fused if possible is difficult. There are two options
  * Supporting NCHW on FusedBatchNorm on the CPU
  * Use the implementation_selector to have a separate implementation on the CPU and GPU. See [this function](https://github.com/tensorflow/tensorflow/blob/e560a277795a3261744e0873f39fb90ad639412c/tensorflow/python/keras/layers/recurrent_v2.py#L569) for an example of this being done in GRU, where we use the cuDNN op only on GPUs. Currently, there is a bug preventing implementation_selector from working with mixed precision in all cases, but this will be fixed.

I recommend the former approach. The fact FusedBatchNorm does not support NCHW on the CPU is a usability issue of the BatchNormalization error. For example, this raises an error, as BatchNormalization chooses the fused implementation even though it is not supported on the CPU:

```
with tf.device('/CPU:0'):
  x = tf.random.normal((2, 3, 4, 5))
  layer = tf.keras.layers.BatchNormalization(axis=1)
  y= layer(x)
```

**Comment by houtoms** [comment]
@reedwm , I added the NCHW support to CPU impl. and removed the `fused` param. Please chek.

**Comment by reedwm** [comment]
Thank you for the FusedBatchNorm fix! Can you split the FusedBatchNorm changes into a separate PR? It's easier to review and will make rollbacks easier if something goes wrong.

**Comment by houtoms** [comment]
Sure. The split PR is https://github.com/tensorflow/tensorflow/pull/32567. Please chek.

**Comment by reedwm** [comment]
Should I take a look again now the batch norm PR is merged?

**Comment by houtoms** [comment]
Yes, I just solved some conflicts. This PR is much simpler now.

File: tensorflow/python/keras/layers/normalization.py [hunk 9]
```diff
@@ -946,6 +947,24 @@ def __init__(self,
 
     self.supports_masking = True
 
+    self.fused = False
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: Make this private: self._fused. And you should initialize this to None instead of False, since it will be updated in build(). Also add a brief comment, such as "Indicates whether a faster fused implementation can be used. This will be set to True or False in build()".

File: tensorflow/python/keras/layers/normalization.py [hunk 10]
```diff
@@ -946,6 +947,24 @@ def __init__(self,
...
+
+  def _raise_if_fused_cannot_be_used(self, ndims):
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: Rename this to `_fused_can_be_used`

File: tensorflow/python/keras/layers/normalization.py [hunk 11]
```diff
@@ -1012,16 +1030,66 @@ def _broadcast(v):
...
+
+      # Compute layer normalization using the batch_normalization function.
+      outputs = nn.batch_normalization(
+          inputs,
+          mean,
+          variance,
+          offset=offset,
+          scale=scale,
+          variance_epsilon=self.epsilon)
+    else:
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: Currently, all the tests will run with the fused implementation, as they all have the axis(es) being the last index(es). Please add a test that uses the non fused implementation.

Also, add a test that asserts the `_fused` attribute is True if the conditions are met

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by houtoms** [comment]
Thx for the suggestion. The changes are committed.

**Review by reedwm — DISMISSED** [review]

**Comment by saberkun** [comment]
Can we make fusion as an option? We see numeric differences.

**Comment by reedwm** [comment]
I discussed with the API owners. There originally was a `fused` parameter but we decided to remove it

**Comment by reedwm** [comment]
@houtoms sorry for the delay, this is failing some internal tests. I'll have an update soon.

**Comment by houtoms** [comment]
Sure. Thanks for the update.

**Comment by reedwm** [comment]
Sorry for the delay. I added some tests in d38aec73faf44e347662ef7639c1920f11901b02 that fail with this PR. I think if the PR passes those tests, it will also pass the internal tests that were previously failing. Can you please fix the PR? Free feel to increase the rtol/atol slightly if necessary, but right now they need to be significantly increased.

My guess is that this is failing due to the fact [fused_batch_norm sets a minimum value of epsilon here](https://github.com/tensorflow/tensorflow/blob/00fad90125b18b80fe054de1055770cfb8fe4ba3/tensorflow/python/ops/nn_impl.py#L1300). Try only using the fused version if epsilon is at least 1.001e-5. I plan on removing the requirement that epsilon is at least 1.001e-5, as cuDNN removed the requirement, but I have to run extensive internal tests first so it might take a while.

File: tensorflow/python/keras/layers/normalization.py [hunk 12]
```diff
@@ -981,6 +982,31 @@ def __init__(self,
...
+    The self.axis is assumed to have no duplicates.
+    """
+    axis = sorted(self.axis)
+    can_use_fused = False
+
+    if axis[-1] == ndims-1 and axis[-1] - axis[0] == len(axis) - 1:
+      can_use_fused = True
+
+    # Fused layer norm may have numeric issues when epsilon < 1.001e-5 (see
+    # cudnn.h). Also, the gamma dtype needs to be float32 when using fused
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py]: cudnn.h no longer has a requirement on the value of epsilon. Instead, say "fused_batch_norm will silently raise epsilon to be at least 1.001e-5, so we cannot used the fused version if epsilon is below that value."

Also, instead of checking gamma.dtype, just check if self.dtype != 'float32'. Note self.dtype is the variable dtype. Then you can also remove the float64 check in call(). You can update the comment, saying "Also, the variable dtype must be float32, as fused_batch_norm only supports float32 variables.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by reedwm** [comment]
Also, when you make the changes, please add a comment saying "PTAL". Then I'll get an email notification so I know I should continue the review.

**Comment by houtoms** [comment]
@reedwm PTAL

**Review by reedwm — APPROVED** [review]
